pysmt/optimization/optimizer.py

```diff
@@ -87,6 +89,55 @@ def _comparation_functions(self, objective_formula):
         else:
             raise PysmtValueError("Invalid optimization function type: %s" % otype)
 
+    def _get_symbol_type(self, objective_formula):
+        otype = self.environment.stc.get_type(objective_formula)
+        if otype.is_int_type():
+            return INT
+        elif otype.is_real_type():
+            return REAL
+        elif otype.is_bv_type():
+            return BVType(otype.width)
+        else:
+            raise PysmtValueError("Invalid optimization function type: %s" % otype)
+
+    def _get_or(self, objective_formula):
+        otype = self.environment.stc.get_type(objective_formula)
+        mgr = self.environment.formula_manager
+        if otype.is_int_type():
+            return mgr.Or
+        elif otype.is_real_type():
+            return mgr.Or
+        elif otype.is_bv_type():
+            return mgr.BVOr
+        else:
+            raise PysmtValueError("Invalid optimization function type: %s" % otype)
+
+
+    def _get_le(self, objective_formula):
```

Comment:
You already have this in the mixins, the reason why you added it here is because you added the push/pop code in the wrong place